### PR TITLE
Optimization: do not allocate temporary objects in MatrixUtils.transformRect

### PR DESCRIPTION
**Pull request by yjbanov**
## Description

Optimize the performance of `MatrixUtils.transformRect`. This function is called 200-300 times per frame when accessibility is enabled in the material demo page in Flutter Gallery. On Nexus 5X each call costs 2.7 microseconds, which is ~0.8ms of frame time.

The optimization performs the transformation as a single matrix multiplication, and it uses a statically allocated temporary matrix to store intermediate results. This removes 4 `Offset` objects, 4 `Vector3` objects, 4 `Float64List` objects. It replaces 4 separate Matrix x Vector multiplications with one Matrix x Matrix multiplication.

Results:

- Dart AOT (Nexus 5X): 1.42x speed-up (2.7µs => 1.89µs average)
- dart2js (Nexus 5X):  3.09x speed-up (1.52µs => 0.49µs average)
- dart2js (Xeon): 3.84x speed-up (0.62µs => 0.16µs average)

## Related Issues

No issues have been filed for this, but we do observe slowness in Flutter for web when accessibility is enabled. `MatrixUtils.transformRect` shows up in our performance profiles.

## Tests

I added the following tests:

- Added new tests in `matrix_utils_test.dart`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [X] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by liyuqian** [comment]
Let's add a benchmark to the device lab?

File: packages/flutter/lib/src/painting/matrix_utils.dart [hunk 1]
```diff
@@ -129,23 +129,98 @@ class MatrixUtils {
...
-    final Offset point2 = transformPoint(transform, rect.topRight);
-    final Offset point3 = transformPoint(transform, rect.bottomLeft);
-    final Offset point4 = transformPoint(transform, rect.bottomRight);
-    return Rect.fromLTRB(
-        _min4(point1.dx, point2.dx, point3.dx, point4.dx),
-        _min4(point1.dy, point2.dy, point3.dy, point4.dy),
-        _max4(point1.dx, point2.dx, point3.dx, point4.dx),
-        _max4(point1.dy, point2.dy, point3.dy, point4.dy),
-    );
+    // Put the coordinates of the four corners of the rect in a matrix in a
```
Comment (jim-flar) [inline packages/flutter/lib/src/painting/matrix_utils.dart]: I would do the following:

bx, by = transform(0, 0)
wx, wy = deltaTransform(0, w)
hx, hy = deltaTransform(h, 0)

this should be far fewer computations than either a matrix multiply or 4 vector multiplies.

The 4 points are now (bx, by), (bx+wx, by+wy), (bx+wx+hx, by+wy+hy), (bx+hx, by+hy)

You don't need to do all those additions to compute the bounding box, though, just do some min/maxing on the wxy,hxy values.

**Comment by yjbanov** [comment]
@liyuqian 

> Let's add a benchmark to the device lab?

Done.

File: packages/flutter/lib/src/painting/matrix_utils.dart (hunk 1 above)
Comment (yjbanov) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: What are `rectx`, `recty` and `rectw`?

Is `transform(rectx, recty)` the same as `transform.transform(Vector4(rectx, recty, 0, 1))`?

I'm also not familiar with `deltaTransform`. What does it do?

Comment (jim-flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: The above optimization definitely works in the absence of perspective which I'm guessing is 99.999% of the operations we do, so even if we just test for a perspective transform and punt to the old code, the delta transform code will be much faster when they aren't performing 3D transforms.

Comment (jim-flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: rectx = left
recty = top
rectw = right-left
recth = bottom-top

And yes, transform(x,y) is my shorthand for transform(x, y, 0, 1).

Basically in the absence of perspective the 4 points form a parallelogram (they start as a rectangle, but after transformation can be an arbitrary parallelogram) which can be transformed with 1 full transform and 2 delta transforms to save a lot of calculations.

**Comment by jim-flar** [comment]
I have an alternate fix which is even faster (more than 2x faster than the suggested fix for common cases and just under 2x faster for perspective transforms), but I've only run the benchmark on Android devices so far.  We can talk about how to run the benchmarks on Chrome when Yegor has some free time to run me through the process.

The results I see on two Motorola devices are:

// On Moto E2:
// I/flutter (15259): MatrixUtils.transformRectClassic: 3.2 µs per iteration
// I/flutter (15259): MatrixUtils.transformRectYegor: 2.3 µs per iteration
// I/flutter (15259): MatrixUtils.transformRectFlar: 1.0 µs per iteration
// I/flutter (15259): MatrixUtils.pTransformRectClassic: 3.2 µs per iteration
// I/flutter (15259): MatrixUtils.pTransformRectYegor: 2.3 µs per iteration
// I/flutter (15259): MatrixUtils.pTransformRectFlar: 1.4 µs per iteration

// On Moto G4:
// I/flutter (20474): MatrixUtils.transformRectClassic: 1.9 µs per iteration
// I/flutter (20474): MatrixUtils.transformRectYegor: 1.2 µs per iteration
// I/flutter (20474): MatrixUtils.transformRectFlar: 0.5 µs per iteration
// I/flutter (20474): MatrixUtils.pTransformRectClassic: 1.8 µs per iteration
// I/flutter (20474): MatrixUtils.pTransformRectYegor: 1.3 µs per iteration
// I/flutter (20474): MatrixUtils.pTransformRectFlar: 0.7 µs per iteration

File: dev/benchmarks/microbenchmarks/lib/geometry/matrix_utils_transform_rect_bench.dart [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+const int _kNumIters = 10000;
+
+void main() {
+  assert(false, "Don't run benchmarks in checked mode! Use 'flutter run --release'.");
+  print('MatrixUtils.transformRect benchmark...');
+
+  final Matrix4 transform = Matrix4.zero()
+    ..scale(1.2, 1.3, 1.0)
+    ..rotateZ(0.1);
```
Comment (jim-flar) [inline dev/benchmarks/microbenchmarks/lib/geometry/matrix_utils_transform_rect_bench.dart]: This produces a degenerate matrix which consists of all 0s and results in NaN for all coordinates.  The .zero() call should probably be replaced by .identity().

**Comment by goderbauer** [comment]
(PR triage): Any plans to came back to this PR?

**Comment by flar** [comment]
I'm preparing a new PR, but holding it until I can chat with Yegor when he returns.

**Comment by yjbanov** [comment]
Update: I'm back and ready to resume this work. However, @flar may have a more optimal solution, so I may close this PR after catching up with him.

**Comment by yjbanov** [comment]
Closing in favor of https://github.com/flutter/flutter/pull/36396

**Pull request closed without merging**